File: crates/kuben-api/src/routes/members.rs
//! Organization members and roles (scenario 4).
//!
//! Rules: nobody grants a role above their own; only owners touch owners;
//! the last owner can be neither demoted nor removed; removing a member
//! revokes their sessions and tokens at once.

use axum::{
    Json,
    extract::{Path, State},
    http::StatusCode,
};
use base64::{Engine as _, engine::general_purpose::URL_SAFE_NO_PAD};
use kuben_core::{
    Error,
    authz::ScopeChain,
    ids::{OrgId, UserId},
    model::Member,
    perm::{Perm, Role},
};
use serde::{Deserialize, Serialize};
use tokio::task::spawn_blocking;
use utoipa::ToSchema;

use super::validate;
use crate::{
    authz::Authz,
    error::{ApiError, ApiResult},
    state::ApiState,
};

fn default_role() -> String {
    "developer".into()
}

#[derive(Debug, Serialize, ToSchema)]
pub struct MemberDto {
    pub id: String,
    pub email: String,
    pub display_name: Option<String>,
    pub role: String,
    /// Invited and has not replaced the temporary password yet.
    pub must_change_password: bool,
    pub active: bool,
}

impl From<&Member> for MemberDto {
    fn from(m: &Member) -> Self {
        Self {
            id: m.user.id.to_string(),
            email: m.user.email.clone(),
            display_name: m.user.display_name.clone(),
            role: m.role.to_string(),
            must_change_password: m.user.must_change_password,
            active: m.user.is_active,
        }
    }
}

#[derive(Debug, Deserialize, ToSchema)]
pub struct InviteMember {
    #[schema(example = "carol@example.com")]
    pub email: String,
    pub display_name: Option<String>,
    /// `viewer`, `developer`, `admin` or `owner`.
    #[serde(default = "default_role")]
    pub role: String,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct InvitedMember {
    pub member: MemberDto,
    /// Temporary password for a newly created account. Shown once; it must be
    /// replaced at the first login. `null` when the user already existed.
    pub temporary_password: Option<String>,
}

#[derive(Debug, Deserialize, ToSchema)]
pub struct UpdateMember {
    pub role: String,
}

fn org_of(authz: &Authz) -> Result<OrgId, ApiError> {
    authz.org_ids().first().copied().ok_or(ApiError(Error::Forbidden))
}

async fn find_member(state: &ApiState, org: OrgId, id: &str) -> ApiResult<Member> {
    let not_found = || ApiError(Error::NotFound(format!("member `{id}`")));
    let user: UserId = id.parse().map_err(|_| not_found())?;
    state
        .store
        .list_members(org)
        .await?
        .into_iter()
        .find(|m| m.user.id == user)
        .ok_or_else(not_found)
}

/// Caller's own org role, required for every change.
fn caller_role(authz: &Authz, org: OrgId) -> Result<Role, ApiError> {
    authz.org_role(org).ok_or(ApiError(Error::Forbidden))
}

/// Members of the caller's organization.
#[utoipa::path(
    get,
    path = "/members", operation_id = "listMembers",
    tag = "members",
    responses((status = 200, body = Vec<MemberDto>), (status = 403, body = crate::error::Problem))
)]
pub async fn list(State(state): State<ApiState>, authz: Authz) -> ApiResult<Json<Vec<MemberDto>>> {
    let org = org_of(&authz)?;
    let _proof = authz.require(&state, Perm::OrgRead, &ScopeChain::org(org))?;
    let members = state.store.list_members(org).await?;
    Ok(Json(members.iter().map(MemberDto::from).collect()))
}

/// Invite a member. New accounts get a one-time temporary password.
#[utoipa::path(
    post,
    path = "/members", operation_id = "inviteMember",
    tag = "members",
    request_body = InviteMember,
    responses(
        (status = 201, body = InvitedMember),
        (status = 403, body = crate::error::Problem),
        (status = 409, body = crate::error::Problem),
        (status = 422, body = crate::error::Problem),
    )
)]
pub async fn invite(
    State(state): State<ApiState>,
    authz: Authz,
    Json(body): Json<InviteMember>,
) -> ApiResult<(StatusCode, Json<InvitedMember>)> {
    authz.forbid_token()?;
    let org = org_of(&authz)?;
    let _proof = authz.require(&state, Perm::UserAdmin, &ScopeChain::org(org))?;
    let role: Role = body.role.parse()?;
    if role.rank() > caller_role(&authz, org)?.rank() {
        return Err(Error::Forbidden.into());
    }
    let email = body.email.trim().to_ascii_lowercase();
    validate::email(&email)?;

    let (user, temporary_password) = if let Some(existing) = state.store.find_user_by_email(&email).await? {
        if state
            .store
            .list_members(org)
            .await?
            .iter()
            .any(|m| m.user.id == existing.user.id)
        {
            return Err(Error::Conflict(format!("`{email}` is already a member")).into());
        }
        (existing.user, None)
    } else {
        let bytes: [u8; 18] = rand::random();
        let password = URL_SAFE_NO_PAD.encode(bytes);
        let hasher = state.hasher.clone();
        let to_hash = password.clone();
        let hash = spawn_blocking(move || hasher.hash(&to_hash))
            .await
            .map_err(Error::internal)?
            .map_err(Error::internal)?;
        let display_name = body
            .display_name
            .as_deref()
            .map(str::trim)
            .filter(|n| !n.is_empty());
        let user = state
            .store
            .create_invited_user(&email, display_name, Some(&hash))
            .await?;
        (user, Some(password))
    };
    state.store.add_membership(org, user.id).await?;
    state.store.bind_org_role(org, user.id, role).await?;
    let member = Member { user, role };
    Ok((
        StatusCode::CREATED,
        Json(InvitedMember {
            member: MemberDto::from(&member),
            temporary_password,
        }),
    ))
}

/// Change a member's role.
#[utoipa::path(
    patch,
    path = "/members/{member}", operation_id = "updateMember",
    tag = "members",
    params(("member" = String, Path, description = "User id")),
    request_body = UpdateMember,
    responses(
        (status = 200, body = MemberDto),
        (status = 403, body = crate::error::Problem),
        (status = 409, body = crate::error::Problem),
    )
)]
pub async fn update(
    State(state): State<ApiState>,
    authz: Authz,
    Path(member): Path<String>,
    Json(body): Json<UpdateMember>,
) -> ApiResult<Json<MemberDto>> {
    authz.forbid_token()?;
    let org = org_of(&authz)?;
    let _proof = authz.require(&state, Perm::UserAdmin, &ScopeChain::org(org))?;
    let role: Role = body.role.parse()?;
    let target = find_member(&state, org, &member).await?;
    if target.user.id == authz.current.user.id {
        return Err(Error::Conflict("you cannot change your own role".into()).into());
    }
    let caller = caller_role(&authz, org)?;
    let touches_owner = target.role == Role::Owner || role == Role::Owner;
    if (touches_owner && caller != Role::Owner) || role.rank() > caller.rank() {
        return Err(Error::Forbidden.into());
    }
    if target.role == Role::Owner && role != Role::Owner && state.store.count_owners(org).await? <= 1 {
        return Err(Error::Conflict("the last owner cannot be demoted".into()).into());
    }
    state.store.set_org_role(org, target.user.id, role).await?;
    Ok(Json(MemberDto::from(&Member {
        user: target.user,
        role,
    })))
}

/// Remove a member: bindings, sessions and tokens are revoked immediately.
#[utoipa::path(
    delete,
    path = "/members/{member}", operation_id = "removeMember",
    tag = "members",
    params(("member" = String, Path, description = "User id")),
    responses(
        (status = 204, description = "Removed"),
        (status = 403, body = crate::error::Problem),
        (status = 409, body = crate::error::Problem),
    )
)]
pub async fn remove(
    State(state): State<ApiState>,
    authz: Authz,
    Path(member): Path<String>,
) -> ApiResult<StatusCode> {
    authz.forbid_token()?;
    let org = org_of(&authz)?;
    let _proof = authz.require(&state, Perm::UserAdmin, &ScopeChain::org(org))?;
    let target = find_member(&state, org, &member).await?;
    if target.user.id == authz.current.user.id {
        return Err(Error::Conflict("you cannot remove yourself".into()).into());
    }
    if target.role == Role::Owner {
        if caller_role(&authz, org)? != Role::Owner {
            return Err(Error::Forbidden.into());
        }
        if state.store.count_owners(org).await? <= 1 {
            return Err(Error::Conflict("the last owner cannot be removed".into()).into());
        }
    }
    state.store.remove_member(org, target.user.id).await?;
    state.store.revoke_all_sessions(target.user.id).await?;
    state.store.revoke_user_tokens(org, target.user.id).await?;
    state.session_cache.invalidate_all();
    Ok(StatusCode::NO_CONTENT)
}

File: go/hub/internal/api/members.go
package api

import (
	"context"
	"crypto/rand"
	"encoding/base64"
	"strings"
	"unicode"

	"github.com/Teamtem-dev/kuben/go/hub/internal/api/access"
	"github.com/Teamtem-dev/kuben/go/hub/internal/api/gen"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/ascii"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/authz"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/ids"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/kerr"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/model"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/opt"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/perm"
)

// Organization members and roles (routes/members.rs, scenario 4). Nobody
// grants a role above their own; only owners touch owners; the last owner
// can be neither demoted nor removed; removing a member revokes their
// sessions and tokens at once.

// memberDto is MemberDto::from.
func memberDto(m model.Member) gen.MemberDto {
	return gen.MemberDto{
		ID:                 m.User.ID.String(),
		Email:              m.User.Email,
		DisplayName:        optNilString(m.User.DisplayName),
		Role:               m.Role.String(),
		MustChangePassword: m.User.MustChangePassword,
		Active:             m.User.IsActive,
	}
}

func memberDtos(members []model.Member) []gen.MemberDto {
	out := make([]gen.MemberDto, 0, len(members))
	for _, m := range members {
		out = append(out, memberDto(m))
	}
	return out
}

// ValidEmail is validate::email: at most 254 bytes, no whitespace, a
// non-empty local part and a domain with a dot and no second `@`.
func ValidEmail(value string) error {
	local, domain, found := strings.Cut(value, "@")
	ok := len(value) <= 254 && !strings.ContainsFunc(value, unicode.IsSpace) &&
		found && local != "" && strings.Contains(domain, ".") && !strings.Contains(domain, "@")
	if !ok {
		return kerr.New(kerr.Validation, "`%s` is not a valid email address", value)
	}
	return nil
}

// orgOf is the organization member management acts on: the caller's
// first.
func orgOf(a access.Access) (ids.OrgID, error) {
	orgs := a.OrgIDs()
	if len(orgs) == 0 {
		return ids.OrgID{}, kerr.ErrForbidden
	}
	return orgs[0], nil
}

// callerRole is the caller's own org role, required for every change.
func callerRole(a access.Access, org ids.OrgID) (perm.Role, error) {
	role, ok := a.OrgRole(org).Get()
	if !ok {
		return "", kerr.ErrForbidden
	}
	return role, nil
}

// memberAdmin is the access, organization and proof of user-admin every
// member change starts with; tokens never manage members.
func (s *Server) memberAdmin(ctx context.Context) (access.Access, ids.OrgID, error) {
	a, err := s.access(ctx)
	if err != nil {
		return access.Access{}, ids.OrgID{}, err
	}
	if err := a.ForbidToken(); err != nil {
		return access.Access{}, ids.OrgID{}, err //nolint:wrapcheck // a kerr already
	}
	org, err := orgOf(a)
	if err != nil {
		return access.Access{}, ids.OrgID{}, err
	}
	if _, err := a.Require(perm.UserAdmin, authz.OrgChain(org)); err != nil {
		return access.Access{}, ids.OrgID{}, err //nolint:wrapcheck // a kerr already
	}
	return a, org, nil
}

func (s *Server) findMember(ctx context.Context, org ids.OrgID, id string) (model.Member, error) {
	notFound := kerr.New(kerr.NotFound, "member `%s`", id)
	user, err := ids.Parse[ids.User](id)
	if err != nil {
		return model.Member{}, notFound
	}
	members, err := s.deps.Store.ListMembers(ctx, org)
	if err != nil {
		return model.Member{}, err //nolint:wrapcheck // a store error, answered as internal
	}
	for _, m := range members {
		if m.User.ID == user {
			return m, nil
		}
	}
	return model.Member{}, notFound
}

// ListMembers is the members of the caller's organization.
func (s *Server) ListMembers(ctx context.Context) (gen.ListMembersRes, error) {
	a, err := s.access(ctx)
	if err != nil {
		return nil, err
	}
	org, err := orgOf(a)
	if err != nil {
		return nil, err
	}
	if _, err := a.Require(perm.OrgRead, authz.OrgChain(org)); err != nil {
		return nil, err //nolint:wrapcheck // a kerr already
	}
	members, err := s.deps.Store.ListMembers(ctx, org)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	out := gen.ListMembersOKApplicationJSON(memberDtos(members))
	return &out, nil
}

// InviteMember invites a member. New accounts get a one-time temporary
// password.
func (s *Server) InviteMember(ctx context.Context, req *gen.InviteMember) (gen.InviteMemberRes, error) {
	a, org, err := s.memberAdmin(ctx)
	if err != nil {
		return nil, err
	}
	role, err := perm.ParseRole(req.Role.Or(string(perm.Developer)))
	if err != nil {
		return nil, err //nolint:wrapcheck // a kerr already
	}
	caller, err := callerRole(a, org)
	if err != nil {
		return nil, err
	}
	if role.Rank() > caller.Rank() {
		return nil, kerr.ErrForbidden
	}
	email := ascii.Lower(strings.TrimSpace(req.Email))
	if err := ValidEmail(email); err != nil {
		return nil, err
	}
	user, temporary, err := s.inviteeAccount(ctx, org, email, req.DisplayName)
	if err != nil {
		return nil, err
	}
	if err := s.deps.Store.AddMembership(ctx, org, user.ID); err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	if err := s.deps.Store.BindOrgRole(ctx, org, user.ID, role); err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	return &gen.InvitedMember{
		Member:            memberDto(model.Member{User: user, Role: role}),
		TemporaryPassword: optNilString(temporary),
	}, nil
}

// inviteeAccount is the existing account of email (not yet a member of
// org), or a new one with a temporary password it must replace at the
// first login.
func (s *Server) inviteeAccount(
	ctx context.Context, org ids.OrgID, email string, displayName gen.OptNilString,
) (model.User, opt.Val[string], error) {
	existing, found, err := s.deps.Store.FindUserByEmail(ctx, email)
	if err != nil {
		return model.User{}, opt.None[string](), err //nolint:wrapcheck // a store error, answered as internal
	}
	if found {
		members, err := s.deps.Store.ListMembers(ctx, org)
		if err != nil {
			return model.User{}, opt.None[string](), err //nolint:wrapcheck // a store error, answered as internal
		}
		for _, m := range members {
			if m.User.ID == existing.User.ID {
				return model.User{}, opt.None[string](), kerr.New(kerr.Conflict, "`%s` is already a member", email)
			}
		}
		return existing.User, opt.None[string](), nil
	}
	var b [18]byte
	_, _ = rand.Read(b[:]) //nolint:errcheck // crypto/rand.Read never fails (Go ≥ 1.24)
	password := base64.RawURLEncoding.EncodeToString(b[:])
	var hash string
	var hashErr error
	if err := s.withPermit(ctx, func() { hash, hashErr = s.deps.Hasher.Hash(password) }); err != nil {
		return model.User{}, opt.None[string](), err
	}
	if hashErr != nil {
		return model.User{}, opt.None[string](), kerr.Wrap(hashErr, "hash the temporary password")
	}
	name := opt.None[string]()
	if n, ok := displayName.Get(); ok {
		if n = strings.TrimSpace(n); n != "" {
			name = opt.Some(n)
		}
	}
	user, err := s.deps.Store.CreateInvitedUser(ctx, email, name, opt.Some(hash))
	if err != nil {
		return model.User{}, opt.None[string](), err //nolint:wrapcheck // a store error, answered as internal
	}
	return user, opt.Some(password), nil
}

// UpdateMember changes a member's role.
func (s *Server) UpdateMember(ctx context.Context, req *gen.UpdateMember, params gen.UpdateMemberParams) (gen.UpdateMemberRes, error) {
	a, org, err := s.memberAdmin(ctx)
	if err != nil {
		return nil, err
	}
	role, err := perm.ParseRole(req.Role)
	if err != nil {
		return nil, err //nolint:wrapcheck // a kerr already
	}
	target, err := s.findMember(ctx, org, params.Member)
	if err != nil {
		return nil, err
	}
	if target.User.ID == a.Current.User.ID {
		return nil, kerr.New(kerr.Conflict, "you cannot change your own role")
	}
	caller, err := callerRole(a, org)
	if err != nil {
		return nil, err
	}
	touchesOwner := target.Role == perm.Owner || role == perm.Owner
	if (touchesOwner && caller != perm.Owner) || role.Rank() > caller.Rank() {
		return nil, kerr.ErrForbidden
	}
	if target.Role == perm.Owner && role != perm.Owner {
		owners, err := s.deps.Store.CountOwners(ctx, org)
		if err != nil {
			return nil, err //nolint:wrapcheck // a store error, answered as internal
		}
		if owners <= 1 {
			return nil, kerr.New(kerr.Conflict, "the last owner cannot be demoted")
		}
	}
	if err := s.deps.Store.SetOrgRole(ctx, org, target.User.ID, role); err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	dto := memberDto(model.Member{User: target.User, Role: role})
	return &dto, nil
}

// RemoveMember removes a member: bindings, sessions and tokens are revoked
// immediately.
func (s *Server) RemoveMember(ctx context.Context, params gen.RemoveMemberParams) (gen.RemoveMemberRes, error) {
	a, org, err := s.memberAdmin(ctx)
	if err != nil {
		return nil, err
	}
	target, err := s.findMember(ctx, org, params.Member)
	if err != nil {
		return nil, err
	}
	if target.User.ID == a.Current.User.ID {
		return nil, kerr.New(kerr.Conflict, "you cannot remove yourself")
	}
	if target.Role == perm.Owner {
		caller, err := callerRole(a, org)
		if err != nil {
			return nil, err
		}
		if caller != perm.Owner {
			return nil, kerr.ErrForbidden
		}
		owners, err := s.deps.Store.CountOwners(ctx, org)
		if err != nil {
			return nil, err //nolint:wrapcheck // a store error, answered as internal
		}
		if owners <= 1 {
			return nil, kerr.New(kerr.Conflict, "the last owner cannot be removed")
		}
	}
	// Bindings, membership and the member's tokens of this org go together.
	if err := s.deps.Store.RemoveMember(ctx, org, target.User.ID); err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	if _, err := s.deps.Store.RevokeAllSessions(ctx, target.User.ID); err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	s.sessionCache.Purge()
	return &gen.RemoveMemberNoContent{}, nil
}
